src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+        else {
+            result = ast.getParent().getFirstChild();
+        }
+        return result;
+    }
+
+    /**
+     * Returns list of nodes matching defined line and column number.
+     * @param root {@code DetailAST} root ast
+     * @param lineNumber line number
+     * @param columnNumber column number
+     * @return list of nodes matching defined line and column number
+     */
+    private static List<DetailAST> getMatchingAstElements(DetailAST root, int lineNumber,
+                                                          int columnNumber) {
+        final List<DetailAST> result = new ArrayList<>();
+        DetailAST curNode = root;
+        while (curNode != null && curNode.getLineNo() <= lineNumber) {
+            if (curNode.getLineNo() == lineNumber
+                && curNode.getColumnNo() == columnNumber
+                && curNode.getType() != TokenTypes.IDENT) {
+                result.add(curNode);
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            curNode = toVisit;
+        }
+        return result;
+    }
+
+    /**
+     * Returns relative xpath query for given ast element from root.
+     * @param root {@code DetailAST} root element
+     * @param ast {@code DetailAST} ast element
+     * @return relative xpath query for given ast element from root
+     */
+    private static String getXpathQuery(DetailAST root, DetailAST ast) {
+        final StringBuilder resultBuilder = new StringBuilder(1024);
+        DetailAST cur = ast;
+        while (cur != root) {
+            final StringBuilder curNodeQueryBuilder = new StringBuilder(256);
+            curNodeQueryBuilder.append('/')
+                    .append(TokenUtils.getTokenName(cur.getType()));
+            final DetailAST identAst = cur.findFirstToken(TokenTypes.IDENT);
+            if (identAst != null) {
+                curNodeQueryBuilder.append("[@text='")
+                        .append(identAst.getText())
+                        .append("']");
+            }
+            resultBuilder.insert(0, curNodeQueryBuilder);
+            cur = cur.getParent();
+        }
+        return resultBuilder.toString();
+    }
+
+    /**
+     * Checks if the given ast element has unique {@code TokenTypes} among siblings.
+     * @param ast {@code DetailAST} ast element
+     * @return if the given ast element has unique {@code TokenTypes} among siblings
+     */
+    private static boolean hasAtLeastOneSiblingWithSameTokenType(DetailAST ast) {
+        DetailAST cur = getFirstSibling(ast);
+        boolean result = false;
+        while (cur != null) {
+            if (cur != ast && cur.getType() == ast.getType()) {
```

Comment:
Why is it required to compare ```cur != ast``` ? Just for ```testNotUniqueDetailAst```? If yes, it is incorrect, as you cheat manually creating the tree in test.